src/protections/BanInviteRetraction.ts

```diff
@@ -0,0 +1,160 @@
+    capabilities: CapabilitySet,
+    protectedRoomsSet: ProtectedRoomsSet,
+    private readonly draupnir: Draupnir,
+  ) {
+    super(description, capabilities, protectedRoomsSet, {
+      requiredPermissions: [PowerLevelPermission.Kick],
+    });
+  }
+
+  public retractInvitesForNewUserPolicy(policy: PolicyRule): void {
+    const rooms: StringRoomID[] = [];
+    if (policy.isGlob()) {
+      this.protectedRoomsSet.allProtectedRooms.forEach((room) =>
+        rooms.push(room.toRoomIDOrAlias()),
+      );
+    } else {
+      for (const roomMembership of this.protectedRoomsSet.setRoomMembership
+        .allRooms) {
+        const membership = roomMembership.membershipForUser(
+          policy.entity as StringUserID,
+        );
+        if (membership !== undefined) {
+          rooms.push(roomMembership.room.toRoomIDOrAlias());
+        }
+      }
+    }
+    void Task(
+      redactUserMessagesIn(
+        this.draupnir.client,
+        this.draupnir.managementRoomOutput,
+        policy.entity,
+        rooms,
+      ),
+    );
+  }
+
+  public async handlePolicyChange(
+    revision: PolicyListRevision,
+    changes: PolicyRuleChange[],
+  ): Promise<ActionResult<void>> {
+    const relevantChanges = changes.filter(
+      (change) =>
+        change.changeType === SimpleChangeType.Added &&
+        change.rule.kind === PolicyRuleType.User,
+    );
+
+    // We don't care about the ramifications of adding a new list here.
+    // The set of changes realistically should be fairly low.
+    relevantChanges.forEach((change) => {
+      this.retractInvitesForNewUserPolicy(change.rule);
+    });
+    return Ok(undefined);
+  }
+
+  public async handleMembershipChange(
+    revision: RoomMembershipRevision,
+    changes: MembershipChange[],
+  ): Promise<ActionResult<void>> {
+    const isUserJoiningWithPolicyRequiringRedaction = (
+      change: MembershipChange,
+    ) => {
+      if (
+        change.membershipChangeType === MembershipChangeType.Joined ||
+        change.membershipChangeType === MembershipChangeType.Rejoined
+      ) {
+        const policyRevision =
+          this.protectedRoomsSet.watchedPolicyRooms.currentRevision;
+        const matchingPolicy = policyRevision.findRuleMatchingEntity(
+          change.userID,
+          PolicyRuleType.User,
+          Recommendation.Ban,
+        );
+        return matchingPolicy !== undefined;
+      } else {
+        return false;
+      }
+    };
+    const relevantChanges = changes.filter(
+      isUserJoiningWithPolicyRequiringRedaction,
+    );
+    for (const change of relevantChanges) {
+      void Task(
+        this.retractInvitesFromUserIn(
+          revision.room.toRoomIDOrAlias(),
+          change.userID,
+        ),
+      );
+    }
+    return Ok(undefined);
+  }
+
+  private async retractInvitesFromUserIn(room: StringRoomID, user: StringUserID): Promise<void> {
+    const roomState = await this.draupnir.client.getRoomState(room);
```

Comment:
There's no need to do this. We have access to immediate room state and filters via `this.protectedRoomsSet.setRoomState.getRevisionForRoom`